File: integrate-routes.py
#!/usr/bin/env python3
"""
Smart API Route Integration Script
Converts remaining routes to use error-handler and response-handler
"""

import os
import re
import sys
from pathlib import Path
from typing import Tuple, Optional

class RouteIntegrator:
    def __init__(self, dry_run=False):
        self.dry_run = dry_run
        self.processed = 0
        self.changed = 0
        self.failed = 0
        self.skipped = 0

    def is_already_integrated(self, content: str) -> bool:
        """Check if route already uses error handlers"""
        return '@/lib/utils/error-handler' in content

    def needs_integration(self, content: str) -> bool:
        """Check if route needs integration"""
        return 'NextResponse' in content and not self.is_already_integrated(content)

    def add_imports(self, content: str) -> str:
        """Add error and response handler imports"""
        # If already has imports, skip
        if '@/lib/utils/error-handler' in content:
            return content

        # Remove NextResponse import
        content = re.sub(
            r'import\s*{\s*NextResponse,\s*type\s*NextRequest\s*}\s*from\s*["\']next/server["\'];?',
            'import { type NextRequest } from "next/server";',
            content
        )
        
        # Remove standalone NextResponse import
        content = re.sub(
            r'import\s*{\s*NextResponse\s*}\s*from\s*["\']next/server["\'];?\s*\n',
            '',
            content
        )

        # Find the last import statement
        lines = content.split('\n')
        last_import_idx = -1
        for i, line in enumerate(lines):
            if line.startswith('import '):
                last_import_idx = i

        if last_import_idx >= 0:
            # Insert new imports after last import
            new_imports = [
                'import { AuthenticationError, AuthorizationError, ValidationError, NotFoundError, ConflictError, DatabaseError, ExternalServiceError, InternalServerError } from "@/lib/utils/error-handler";',
                'import { successResponse, errorResponse, createdResponse } from "@/lib/utils/response-handler";'
            ]
            lines.insert(last_import_idx + 1, '')
            for imp in reversed(new_imports):
                lines.insert(last_import_idx + 1, imp)
            content = '\n'.join(lines)

        return content

    def convert_responses(self, content: str) -> str:
        """Convert NextResponse patterns to new handlers"""
        
        # Pattern 1: NextResponse.json({ success: true, data: X }, { status: 200 })
        content = re.sub(
            r'NextResponse\.json\(\s*{\s*(?:success:\s*true,\s*)?data:\s*(\w+)\s*}\s*(?:,\s*{\s*status:\s*200\s*})?\s*\)',
            r'successResponse(\1)',
            content
        )

        # Pattern 2: NextResponse.json({ data: X })
        content = re.sub(
            r'NextResponse\.json\(\s*{\s*data:\s*(\w+)\s*}\s*\)',
            r'successResponse(\1)',
            content
        )

        # Pattern 3: NextResponse.json(X, { status: 201 })
        content = re.sub(
            r'NextResponse\.json\(\s*(\w+)\s*,\s*{\s*status:\s*201\s*}\s*\)',
            r'createdResponse(\1)',
            content
        )

        # Pattern 4: throw new errors instead of returning errors
        content = re.sub(
            r'return\s+NextResponse\.json\(\s*{[^}]*error[^}]*}\s*,\s*{\s*status:\s*40[134]\s*}\s*\);?',
            'throw new AuthenticationError("Unauthorized");',
            content
        )

        return content

    def simplify_catch_blocks(self, content: str) -> str:
        """Simplify catch blocks to use errorResponse"""
        # Replace complex catch blocks with simple errorResponse
        pattern = r'catch\s*\(\s*error\s*\)\s*{[^}]*return\s+NextResponse\.json\([^)]*\);?\s*}'
        replacement = 'catch (error) {\n    return errorResponse(error);\n  }'
        content = re.sub(pattern, replacement, content, flags=re.DOTALL)
        return content

    def process_file(self, file_path: str) -> Tuple[bool, Optional[str]]:
        """Process a single file, return (changed, error_msg)"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                original = f.read()

            if self.is_already_integrated(original):
                self.skipped += 1
                return False, None

            if not self.needs_integration(original):
                self.skipped += 1
                return False, None

            content = original
            content = self.add_imports(content)
            content = self.convert_responses(content)
            content = self.simplify_catch_blocks(content)

            if content != original:
                if not self.dry_run:
                    with open(file_path, 'w', encoding='utf-8') as f:
                        f.write(content)
                self.changed += 1
                return True, None
            
            return False, None

        except Exception as e:
            self.failed += 1
            return False, str(e)

    def find_routes(self, root_dir: str) -> list:
        """Find all route.ts files"""
        routes = []
        for path in Path(root_dir).rglob('route.ts'):
            routes.append(str(path))
        return sorted(routes)

    def run(self, root_dir: str):
        """Run integration on all routes"""
        routes = self.find_routes(root_dir)
        
        print(f"\n[SEARCH] Found {len(routes)} total routes")
        
        to_process = [r for r in routes if self.needs_integration(open(r, encoding='utf-8').read())]
        print(f"[PROCESS] Processing {len(to_process)} routes that need integration\n")

        for route in to_process:
            try:
                changed, error = self.process_file(route)
                rel_path = os.path.relpath(route)
                
                if error:
                    print(f"[ERROR] {rel_path}: {error}")
                elif changed:
                    status = "[DRY RUN]" if self.dry_run else ""
                    print(f"[OK] {status} {rel_path}")
                
                self.processed += 1
            except Exception as e:
                print(f"[ERROR] Error: {e}")
                self.failed += 1

        print(f"\n[SUMMARY]")
        print(f"  Total processed: {self.processed}")
        print(f"  Changed: {self.changed}")
        print(f"  Skipped: {self.skipped}")
        print(f"  Failed: {self.failed}")
        
        if self.dry_run:
            print(f"\n[INFO] Run with --apply to make changes")

if __name__ == "__main__":
    dry_run = "--dry-run" in sys.argv
    apply_mode = "--apply" in sys.argv
    
    root = Path(__file__).parent / "src" / "app" / "api"
    
    integrator = RouteIntegrator(dry_run=dry_run and not apply_mode)
    integrator.run(str(root))

File: integrate-routes.js
#!/usr/bin/env node

/**
 * Automated Route Integration Script
 * Converts all remaining API routes to use error-handler and response-handler
 * 
 * Usage: node integrate-routes.js [--dry-run] [--file <path>]
 */

const fs = require("fs");
const path = require("path");

const args = process.argv.slice(2);
const dryRun = args.includes("--dry-run");
const specificFile = args.find((arg, i) => args[i - 1] === "--file");

const INTEGRATION_PATTERNS = [
  {
    name: "Remove NextResponse import",
    find: /import\s*{\s*NextResponse,\s*type\s*NextRequest\s*}\s*from\s*["']next\/server["'];?/,
    replace: 'import { type NextRequest } from "next/server";',
  },
  {
    name: "Remove standalone NextResponse import",
    find: /import\s*{\s*NextResponse\s*}\s*from\s*["']next\/server["'];?/,
    replace: "",
  },
  {
    name: "Add error handler imports (if not present)",
    find: /^(import\s+{.*?}\s+from\s+["']@\/lib\/config\/auth["'];)/m,
    replace:
      '$1\nimport { AuthenticationError, AuthorizationError, ValidationError, NotFoundError, ConflictError, DatabaseError, ExternalServiceError, InternalServerError } from "@/lib/utils/error-handler";\nimport { successResponse, errorResponse, createdResponse } from "@/lib/utils/response-handler";',
  },
  {
    name: "Convert NextResponse.json() errors to thrown errors",
    find: /return\s+NextResponse\.json\(\s*{.*?error:.*?},\s*{?\s*status:\s*401\s*}\s*\);?/gs,
    replace: 'throw new AuthenticationError("Unauthorized");',
  },
  {
    name: "Convert NextResponse.json() with success to successResponse",
    find: /return\s+NextResponse\.json\(\s*{\s*success:\s*true,\s*data:\s*([^}]+)}\s*(?:,\s*{.*?status.*?})?\s*\);?/g,
    replace: "return successResponse($1);",
  },
  {
    name: "Convert catch block NextResponse errors",
    find: /catch\s*\(\s*error\s*\)\s*{\s*(?:console\.error\([^)]*\);)?\s*return\s+NextResponse\.json\([^)]*\);?\s*}/gs,
    replace: "catch (error) {\n    return errorResponse(error);\n  }",
  },
];

function shouldProcessFile(filePath) {
  // Skip if already integrated
  const content = fs.readFileSync(filePath, "utf-8");
  if (
    content.includes(
      'from "@/lib/utils/error-handler"'
    )
  ) {
    return false;
  }
  return content.includes("NextResponse");
}

function processFile(filePath) {
  let content = fs.readFileSync(filePath, "utf-8");
  const originalContent = content;

  // Step 1: Fix imports
  if (!content.includes('from "@/lib/utils/error-handler"')) {
    // Remove NextResponse import
    content = content.replace(
      /import\s*{\s*NextResponse,\s*type\s*NextRequest\s*}\s*from\s*["']next\/server["'];?/,
      'import { type NextRequest } from "next/server";'
    );

    // If still has NextResponse, remove it
    content = content.replace(
      /import\s*{\s*NextResponse\s*}\s*from\s*["']next\/server["'];?/,
      ""
    );

    // Add error/response handlers after auth import
    if (content.includes('from "@/lib/config/auth"')) {
      content = content.replace(
        /(import.*?from\s*["']@\/lib\/config\/auth["'];)/,
        '$1\nimport { AuthenticationError, AuthorizationError, ValidationError, NotFoundError, ConflictError, DatabaseError, ExternalServiceError, InternalServerError } from "@/lib/utils/error-handler";\nimport { successResponse, errorResponse, createdResponse } from "@/lib/utils/response-handler";'
      );
    } else {
      // Add at top of file after all imports
      const lastImportMatch = content.lastIndexOf("import ");
      const lastImportEnd = content.indexOf(";", lastImportMatch) + 1;
      content =
        content.slice(0, lastImportEnd) +
        '\nimport { AuthenticationError, AuthorizationError, ValidationError, NotFoundError, ConflictError, DatabaseError, ExternalServiceError, InternalServerError } from "@/lib/utils/error-handler";\nimport { successResponse, errorResponse, createdResponse } from "@/lib/utils/response-handler";' +
        content.slice(lastImportEnd);
    }
  }

  // Step 2: Replace NextResponse.json() patterns with new handlers
  // This is complex - do simple replacements
  content = content.replace(
    /return\s+NextResponse\.json\(\s*{\s*(?:success:\s*true,\s*)?data:\s*([^}]+)\s*}\s*(?:,\s*{\s*status:\s*200\s*})?\s*\);?/g,
    "return successResponse($1);"
  );

  content = content.replace(
    /return\s+NextResponse\.json\(\s*{[^}]*success:\s*true[^}]*data:\s*([^}]+)[^}]*}\s*\);?/g,
    "return successResponse($1);"
  );

  content = content.replace(
    /return\s+NextResponse\.json\(\s*([^,}]+(?:{[^}]*})?)[^)]*,\s*{\s*status:\s*201\s*}\s*\);?/g,
    "return createdResponse($1);"
  );

  // Replace error responses
  content = content.replace(
    /return\s+NextResponse\.json\(\s*{[^}]*(?:error|message):[^}]*}\s*,\s*{\s*status:\s*40[134]\s*}\s*\);?/g,
    "throw new AuthenticationError('Unauthorized');"
  );

  // Simplify catch blocks
  content = content.replace(
    /catch\s*\(\s*error\s*\)\s*{\s*(?:console\.error\([^)]*\);)?\s*return\s+NextResponse\.json\([^;]*;\s*}/gs,
    "catch (error) {\n    return errorResponse(error);\n  }"
  );

  return { originalContent, newContent: content, changed: content !== originalContent };
}

function findRoutes(dir) {
  let routes = [];
  const items = fs.readdirSync(dir, { withFileTypes: true });

  for (const item of items) {
    const fullPath = path.join(dir, item.name);
    if (item.isDirectory()) {
      routes = routes.concat(findRoutes(fullPath));
    } else if (item.name === "route.ts") {
      routes.push(fullPath);
    }
  }

  return routes;
}

// Main execution
const apiDir = path.join(__dirname, "src", "app", "api");
let allRoutes = findRoutes(apiDir);

// Filter routes that need integration
allRoutes = allRoutes.filter(
  (route) => shouldProcessFile(route) && (!specificFile || route === specificFile)
);

console.log(
  `\n🔍 Found ${allRoutes.length} routes that need integration\n`
);

let processedCount = 0;
let changedCount = 0;
let failedCount = 0;

allRoutes.forEach((routePath) => {
  try {
    const result = processFile(routePath);

    if (result.changed) {
      changedCount++;
      const relativePath = path.relative(process.cwd(), routePath);

      if (dryRun) {
        console.log(`✓ [DRY RUN] Would update: ${relativePath}`);
      } else {
        fs.writeFileSync(routePath, result.newContent, "utf-8");
        console.log(`✓ Updated: ${relativePath}`);
      }
    }

    processedCount++;
  } catch (error) {
    failedCount++;
    console.error(`✗ Error processing ${routePath}: ${error.message}`);
  }
});

console.log(`\n📊 Summary:`);
console.log(`  Total processed: ${processedCount}`);
console.log(`  Changed: ${changedCount}`);
console.log(`  Failed: ${failedCount}`);

if (dryRun) {
  console.log(`\n💡 Run without --dry-run to apply changes`);
}
